### PR TITLE
fix: toggle language selector button

### LINKED ISSUE #5509: Bug: Language Selector when clicking should close
### Issue

Clicking the "Language Selector" on the Header (for example) only opens the Language Selector Dropdown instead of toggling its visibility.

### Desired Effect

Clicking the Language Selector button should "close" the Dropdown if already open

### PR DESCRIPTION
**Pull request by Nirajn2311**
## Description

Fixes bug which prevented closing of language selector dropdown when clicking the language selector button

## Related Issues

Fixes #5509

### Check List

- [x] I have read the [Contributing Guidelines](https://github.com/nodejs/nodejs.org/blob/main/CONTRIBUTING.md) and made commit messages that follow the guideline.
- [x] I have run `npx turbo lint` to ensure the code follows the style guide. And run `npx turbo lint:fix` to fix the style errors if necessary.
- [x] I have run `npx turbo format` to ensure the code follows the style guide.
- [x] I have run `npx turbo test` to check if all tests are passing, and/or `npx turbo test:snapshot` to update snapshots if I created and/or updated React Components.
- [ ] I've covered new added functionality with unit tests if necessary.

### CODE REVIEW

**Comment by Nirajn2311** [comment]
Hmm that's strange, the first commit should've failed the storybook tests as the snapshots weren't updated but it didn't 🤔.

Looks like Turborepo is not validating the cache properly as for both commits it has used a previous runs results thereby not executing a new run in this PR at all.

**Comment by bmuenzenmeyer** [comment]
> Hmm that's strange, the first commit should've failed the storybook tests as the snapshots weren't updated but it didn't 🤔.

agree - not yet sure why that didn't break the hash. Looking at the turborepo docs makes it sound like it should have as a consequence of https://github.com/nodejs/nodejs.org/blob/main/turbo.json#L14

File: components/Common/LanguageSelector/index.tsx [hunk 1]
```diff
@@ -41,7 +41,7 @@ const LanguageSelector = () => {
   });
 
   return (
-    <>
+    <div ref={ref}>
```
Comment (ovflowd) [inline components/Common/LanguageSelector/index.tsx]: Please do not add a div here. We intentionally changed the Dropdown to contain the `ref` for obvious reasons.

Please undo this change. See details here: https://github.com/nodejs/nodejs.org/pull/5500

Comment (ovflowd) [reply components/Common/LanguageSelector/index.tsx]: (Please also add the ref back to the Dropdown), why did you remove the ref from there?

**Review by ovflowd — CHANGES_REQUESTED** [review]
Against this change. It breaks previous changes that were intentionally done per-design.

**Comment by ovflowd** [comment]
> agree - not yet sure why that didn't break the hash. Looking at the turborepo docs makes it sound like it should have as a consequence of [main/turbo.json#L14](https://github.com/nodejs/nodejs.org/blob/main/turbo.json?rgh-link-date=2023-07-17T03%3A56%3A11Z#L14)

Yes, we had a cache hit (https://github.com/nodejs/nodejs.org/actions/runs/5567221531/jobs/10168922245). Which doesn't make much sense, as it is clearly shown here (https://github.com/nodejs/nodejs.org/blob/main/turbo.json#L14) that any changes to `components/**` should be computed against, and the first commit clearly signs that changes were done there... I wonder if we're doing something wrong with that groupping.

File: components/Common/LanguageSelector/index.tsx (hunk 1 above)
Comment (Nirajn2311) [reply components/Common/LanguageSelector/index.tsx]: > why did you remove the ref from there?

Well with the `ref` being set in the parent `div`, the `Dropdown` didn't need the `ref` as the same functionality could be done without passing it down.

**Comment by Nirajn2311** [comment]
> Against this change. It breaks previous changes that were intentionally done per-design.

I have an alternate solution in mind which would resolve both #5508 and #5509. This way there won't be an extra parent `div` and the `Dropdown` component would have the `ref` passed to it. Can I go ahead and fix both of them in one PR?

File: components/Common/LanguageSelector/index.tsx (hunk 1 above)
Comment (ovflowd) [reply components/Common/LanguageSelector/index.tsx]: I still ask you, please, to revert this change. We don't want a pointless `div` surrounding the Language Selector button. Not to mention this is a regression to the bug mentioned before.

Please also refer to https://github.com/nodejs/nodejs.org/issues/5508

**Comment by ovflowd** [comment]
> I have an alternate solution in mind which would resolve both https://github.com/nodejs/nodejs.org/issues/5508 and https://github.com/nodejs/nodejs.org/issues/5509. This way there won't be an extra parent div and the Dropdown component would have the ref passed to it. Can I go ahead and fix both of them in one PR?

Sorry, no. No changes at all on the DOM tree are required to fix the "toggle language selector" button. You simply set a state that toggles instead of always setting to open, that's it.

Now regarding https://github.com/nodejs/nodejs.org/issues/5508, the path is explicit and I wish that you could follow it, but in another PR. (This PR should **only** fix the toggle not working)

**Comment by ovflowd** [comment]
In other words, I just urge you to not overcomplicate things and change the way how certain components are structured unnecessarily.

**Comment by ovflowd** [comment]
> Hmm that's strange, the first commit should've failed the storybook tests as the snapshots weren't updated but it didn't 🤔.
> 
> Looks like Turborepo is not validating the cache properly as for both commits it has used a previous runs results thereby not executing a new run in this PR at all.

> > Hmm that's strange, the first commit should've failed the storybook tests as the snapshots weren't updated but it didn't 🤔.
> 
> agree - not yet sure why that didn't break the hash. Looking at the turborepo docs makes it sound like it should have as a consequence of [`main`/turbo.json#L14](https://github.com/nodejs/nodejs.org/blob/main/turbo.json?rgh-link-date=2023-07-17T03%3A56%3A11Z#L14)

I've made a PR that attempts to fix that, but not sure if it will actually fix it. https://github.com/nodejs/nodejs.org/pull/5524 crossing fingers. Both of you feel free to give it a try :)

**Comment by Nirajn2311** [comment]
> You simply set a state that toggles instead of always setting to open, that's it.

But thats the thing, the state is being set properly but it's the order in which they're set that breaks it. The `ref` only tracks the `Dropdown` component. So when someone clicks on the language selector button while the dropdown is open, react will first call the `ref` handler(as it's outside the dropdown) which sets the state `showDropdown` to false([code](https://github.com/nodejs/nodejs.org/blob/8225312fe8d26129e40e222ffed93f92fcfd3c41/components/Common/LanguageSelector/index.tsx#L19)) and also the button click handler also gets executed which in turn sets it back to `true` thereby making it look like it never closed. Now this PR moves the `ref` tracker to a parent element which consists of both the button and dropdown so that it won't execute the `ref` handler when the user clicks the button as this time the click event is not outside the relevant elements(button and dropdown).

Another solution would be to give the button it's own reference and pass that to the `useClickOutside` hook so that it can check if the click event is on the dropdown(which is being tracked by the `ref` variable) or the button(the new reference which we create for the button)

**Comment by ovflowd** [comment]
> But thats the thing, the state is being set properly but it's the order in which they're set that breaks it. The ref only tracks the Dropdown component. So when someone clicks on the language selector button while the dropdown is open, react will first call the ref handler(as it's outside the dropdown) which sets the state showDropdown to false([code](https://github.com/nodejs/nodejs.org/blob/8225312fe8d26129e40e222ffed93f92fcfd3c41/components/Common/LanguageSelector/index.tsx#L19)) and also the button click handler also gets executed which in turn sets it back to true thereby making it look like it never closed.

I do understand the issue. I am still insisting that you can solve this easily without changing the way how the components are rendered.

The solution is simply on how you handle the state changes and the callbacks. 👀

**Comment by ovflowd** [comment]
Please don't get me wrong; I'm all in on you being creative and doing your solution. I'm being intentionally **stubborn** here as I believe you can achieve simplistic results without changing the DOM tree.

If you're struggling with seeing that path, let me know, and I'd be more than happy to support you. But I'd appreciate it if you could solve this without creating regressions 🙇 

On that matter, you can follow your gut/experience. And we can review the PR afterwards ;)

---

**Edit:** I forgot to mention that I might also be wrong here; as I said, I'm being stubborn here and a little bit on the "harsh"-side, but I just wanted to make it clear that I appreciate your contributions, @Nirajn2311 and that I'm here to give you as much context as possible 🙇

File: components/Article/Codebox/__snapshots__/index.stories.tsx.snap
Comment (ovflowd) [inline components/Article/Codebox/__snapshots__/index.stories.tsx.snap:1]: BTW if you could revert this file that'd be awesome. It is related to a regression that @araujogui is probably going to fix.

File: components/Common/LanguageSelector/index.tsx (hunk 1 above)
Comment (bmuenzenmeyer) [reply components/Common/LanguageSelector/index.tsx]: > We don't want a pointless `div` surrounding the Language Selector button

this is the biggest argument to me, from a markup purity standpoint

File: components/Article/Codebox/__snapshots__/index.stories.tsx.snap
Comment (Nirajn2311) [reply components/Article/Codebox/__snapshots__/index.stories.tsx.snap:1]: I've reverted the changes for this file.

File: components/Common/LanguageSelector/index.tsx (hunk 1 above)
Comment (Nirajn2311) [reply components/Common/LanguageSelector/index.tsx]: I've updated the changes to one where the DOM isn't changed.

**Comment by Nirajn2311** [comment]
> I'm being intentionally stubborn here as I believe you can achieve simplistic results without changing the DOM tree.

I've updated this PR with the alternate fix which I mentioned above in my comment. This one doesn't modify the DOM tree

**Comment by ovflowd** [comment]
Sorry, this change is still overcomplicating things. And creates another ref that is unnecessary 😅

**Comment by Nirajn2311** [comment]
Hmm. I'm out of ideas 😅 . What was the fix you had in mind?

**Comment by ovflowd** [comment]
> 
Hmm. I'm out of ideas 😅 . What was the fix you had in mind?

By making the button itself `children` of the Dropdown. Which incidentally fixes #5508...

Hence `ref.current && !ref.current.contains(event.target as Node)` will be `false`.

**Comment by Nirajn2311** [comment]
I had tried that initially but forgot to remove the `ref` from the `ul` element and so it dint work 🤦 .

I've updated the PR now. Thanks for the help and reviews.

**Comment by ovflowd** [comment]
@Nirajn2311 you can fix the issue by simply doing: (Also just as a heads-up)


```tsx
  const ref = useClickOutside<HTMLUListElement>(event => {
    if (showDropdown) {
      event.stopImmediatePropagation();

      setShowDropdown(false);
    }
  });
```

**Comment by Nirajn2311** [comment]
> @Nirajn2311 you can fix the issue by simply doing: (Also just as a heads-up)

Do I implement this or is the current fix acceptable?

**Comment by ovflowd** [comment]
So pretty much the issue, as you stated is that the `handler()` is fired because, yes, the button itself is not a node present of the `ref` itself anymore.

Yet, you can simply customise the handler to prevent further propagation of the event to the `onClick` of the button, if you want. (Which is the actual fix IMHO).

Just in general as guidance, adding an extra convoluted `ref` to a hook that should be generic and ignorant towards the existence of a "button" is not a good design practice.

**Comment by ovflowd** [comment]
> Do I implement this or is the current fix acceptable?

Did you test the current fix? Does it work nicely?

Edit: Your fix seemed to have broken the actual behaviour of `useClickOutside` (@Nirajn2311)

File: components/Common/Dropdown/index.tsx [hunk 2]
```diff
@@ -8,10 +8,11 @@ type DropdownProps = {
   items: Array<DropdownItem>;
   shouldShow: boolean;
   styles: CSSProperties;
+  children: React.ReactNode;
```
Comment (ovflowd) [inline components/Common/Dropdown/index.tsx]: Instead of declaring children here, please wrap `DropdownProps = PropsWithChildren<{ .... }>`

**Comment by ovflowd** [comment]
> > Do I implement this or is the current fix acceptable?
> 
> Did you test the current fix? Does it work nicely?
> 
> Edit: Your fix seemed to have broken the actual behaviour of `useClickOutside` (@Nirajn2311)

Reason being that since you've added `{children}` before the `ul` (dropdown) the `onClick` event handler is fired before the one from the `useClickOutside`.

Remember, events are fired in order of precedence of the Node.

**Comment by Nirajn2311** [comment]
```js
  const ref = useClickOutside<HTMLUListElement>(event => {
    if (showDropdown) {
      event.stopImmediatePropagation();

      setShowDropdown(false);
    }
  });
```

This will still have the bug but in a different manner. The dropdown will never open after one round of opening and closing it.

**Comment by ovflowd** [comment]
I forgot to mention the callback should be on `useCallback` otherwise the event listener will be added infinitely. Such as:

```tsx
  const closeDropdownCallback = useCallback(
    (event: MouseEvent | TouchEvent) => {
      if (showDropdown) {
        event.stopPropagation();

        setShowDropdown(false);
      }
    },
    []
  );

  const ref = useClickOutside<HTMLUListElement>(closeDropdownCallback);
```

**Comment by Nirajn2311** [comment]
Still the bug is there. It'll work the first time and that's it, it'll never open up the second time. I've pushed my latest changes so you can check the code.

File: components/Common/Dropdown/index.tsx [hunk 3]
```diff
@@ -41,9 +41,12 @@ const Dropdown = forwardRef<HTMLUListElement, DropdownProps>(
     };
 
     return (
-      <ul ref={ref} className={styles.dropdownList} style={dropdownStyles}>
-        {mappedElements}
-      </ul>
+      <>
+        {children}
```
Comment (ovflowd) [inline components/Common/Dropdown/index.tsx:45]: As I mentioned before, the order of the children... matters.

Comment (ovflowd) [reply components/Common/Dropdown/index.tsx:45]: Try something like:

```tsx
    return (
      <>
        <ul ref={ref} style={dropdownStyles} className={styles.dropdownList}>
          {mappedElements}
        </ul>

        {children}
      </>
    );
```

Comment (ovflowd) [reply components/Common/Dropdown/index.tsx:45]: Hmm, the callback dependencies, might be wrong here. Let me check something.

Comment (ovflowd) [reply components/Common/Dropdown/index.tsx:45]: Somehow the handler is being registered multiple times. (and not being cleaned up)

**Comment by ovflowd** [comment]
Okay, after some debugging, there are 2 different issues:

1. The `click` listener was being registered multiple times and not being cleaned-up
2. The `onClick` event is actually being handled before the `useClickOutside` event, which is the inverse of my hypothesis.

Then the fix for this is:

```tsx
// callback code should be
const ref = useClickOutside<HTMLUListElement>(() => setShowDropdown(false));

// dropdown return statement should be 
return (
    <Dropdown
      items={dropdownItems}
      shouldShow={showDropdown}
      styles={dropdownStyle}
      ref={ref}
    >
      <button
        type="button"
        className={styles.languageSwitch}
        onClick={event => {
          event.stopPropagation();

          setShowDropdown(!showDropdown);
        }}
        aria-expanded={showDropdown}
        aria-label={ariaLabelText}
      >
        <MdOutlineTranslate />
      </button>
    </Dropdown>
  );

// use effect addEventListener should be
document.addEventListener('click', listener);
```

**Comment by ovflowd** [comment]
@Nirajn2311 do these last changes I've shown fix the issue for you?

**Comment by Nirajn2311** [comment]
Nope, now it's gone back to the original bug ie. not closing the dropdown when clicking the button

**Comment by ovflowd** [comment]
> Nope, now it's gone back to the original bug ie. not closing the dropdown when clicking the button

I see. I'm sorry then, I'm taking-over this PR as clearly I'm not sure how to help you out here besides of literally pasting the code 😅 

I'm making another PR so you can compare with your changes and understand the differences :)

PR: https://github.com/nodejs/nodejs.org/pull/5527

**Comment by Nirajn2311** [comment]
Ahhh I see what I was missing, that third argument for `addEventListener`.

**Comment by ovflowd** [comment]
Small details do make a difference heh! Anyhow, @Nirajn2311 I super appreciate your effort here! 🙇 

Thank you for contributing, nevertheless and looking for future contributions!
